ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,343 @@
+		* If **ignored**, do nothing.
+		* Otherwise, the label cannot be normalized.
+1. Repeat until all the input is consumed.
+1. If the buffer is nonempty, emit a final **Text** token.
+
+### Validate
+
+Given a list of **Emoji** and **Text** tokens, determine if the composition is valid and return the **Label Type**.
+
+1. If only **Emoji** tokens:
+	* Return `"Emoji"`
+1. If a single **Text** token and every characters is ASCII (`00..7F`):
+	* `5F (_) LOW LINE` can only occur at the start.
+		* Must match `/^_*[^_]*$/`
+	* The 3rd and 4th characters must not both be `2D (-) HYPHEN-MINUS`.
+		* Must not match `/^..--/`
+	* Return `"ASCII"`
+		* The label is free of **Fenced** and **Combining Mark** characters, and not confusable.
+1. Concatenate all the tokens together.
+	* `5F (_) LOW LINE` can only occur at the start.
+		* Must match `/^_*[^_]$/u`
+	* The first and last characters cannot be **Fenced**.
+	* **Fenced** characters cannot be contiguous.
+1. The first character of every **Text** token must not be a **Combining Mark**.
+1. Concatenate the **Text** tokens together.
+1. Find the first **Group** that contain every text character:
+	* If no group is found, the label cannot be normalized.
+1. If the group is not **CM Whitelisted**:
+	* Apply NFD to the concatenated text characters.
+	* For every contiguous sequence of **NSM** characters:
+		* Each character must be unique.
+		* Number of characters cannot exceed **Maximum NSM** (4).
+1. [Wholes](#wholes) — check if text characters form a confusable.
+1. The label is valid.
+	* Return the name of the group.
+
+### Wholes
+
+A label is whole-script confusable if a similarly-looking valid label can be constructed using one alternative character from a different group.
+
+1. Allocate an empty character buffer.
+1. Start with the set of all groups.
+1. For each unique character:
+	* If the character is **Confused** (a member of a **Whole Confusable**):
+		* Retain groups with **Whole Confusable** characters excluding the **Confusable Extent** of the matching **Confused** character.
+			* The **Confusable Extent** is the fully-connected graph formed from different groups with the same confusable and different confusables of the same group.
+			* This mapping from **Confused** to **Confusable Extent** can be precomputed.
+			* Example: `"o"` **Whole Confusable** 
+				1. `6F (o) LATIN SMALL LETTER O` → *Latin*, *Han*, *Japanese*, and *Korean*
+				1. `3007 (〇) IDEOGRAPHIC NUMBER ZERO` → *Han*, *Japanese*, *Korean*, and *Bopomofo*
+				1. **Confusable Extent** is [`6F`, `3007`] ⊗ [*Latin*, *Han*, *Japanese*, *Korean*, *Bopomofo*]
+		* If no groups remain, the label is not confusable.
+		* Example: `"тӕ" [442 4D5]`
+			1. `"т"` → `442 (т) CYRILLIC SMALL LETTER TE` and `3C4 (τ) GREEK SMALL LETTER TAU`
+				* **ALL** ∩ [*Cyrillic*, *Greek*] → [*Cyrillic*, *Greek*]
+			1. `"ӕ"` → `E6 (æ) LATIN SMALL LETTER AE` and `4D5 (ӕ) CYRILLIC SMALL LIGATURE A IE`

```

Comment:
Is this example confusable, given there's exactly one script?

Reply:
I actually think it's a good example because confusables involve comparing all visually similar labels but valid labels can only be formed from characters in a single group (which may contain one or more scripts.)

In this example, both `т` and `ӕ` are individually confusable but the composition is only valid in Cyrillic making it not confusable.

Or, if Greek had an `ӕ`-like character, then both Cyrillic (`тӕ`) and Greek (`τæ`) exist, however the single-scriptness of the label doesn't prevent it from being confusing.

I'll include another less-esoteric example, like an `0x` vs `0х`.